Add Agentic Memory APIs (#10714)
opensearch-project/documentation-website: _ml-commons-plugin/api/agentic-memory-apis/add-memory.md

---
layout: default
title: Add Agentic Memory
parent: Agentic Memory APIs
grand_parent: ML Commons APIs
nav_order: 10
---

# Add Agentic Memory API
**Introduced 3.2**
{: .label .label-purple }

Use this API to add a memory into a memory container.

Once a memory is created, you'll provide its `memory_id` to other APIs.

## Endpoint

```json
POST /_plugins/_ml/memory_containers/{memory_container_id}/memories
```

## Request body fields

The following table lists the available request fields.

Field | Data type | Required/Optional | Description
:--- | :--- | :--- | :---
`messages` | List | Required | The list of messages, currently limited to one message per request.
`session_id` | String | Optional | The session ID.
`agent_id` | String | Optional | The agent ID.

> `infer` | Boolean | Optional | Whether to infer context from messages.

Reviewer: What is the default for this parameter?

Reviewer 2: Default is true when LLM_MODEL_ID specified, otherwise false.

`tags` | Object | Optional | Custom metadata for the memory.

### The `infer` parameter behavior

- When `infer=true`: LLM extracts facts, response contains only FACT memories
- When `infer=false`: No LLM processing, response contains the stored message

## Example request

```json
POST /_plugins/_ml/memory_containers/SdjmmpgBOh0h20Y9kWuN/memories
{
    "messages": [
        {"role": "assistant", "content": "Machine learning is a subset of artificial intelligence"}
    ],
    "session_id": "sess_789",
    "agent_id": "agent_123",
    "tags": {
        "topic": "personal info"
    }
}
```
{% include copy-curl.html %}

## Example response

```json
{
    "results": [
        {
            "id": "T9jtmpgBOh0h20Y91WtZ",
            "text": "Machine learning is a subset of artificial intelligence",

>             "event": "ADD"

Reviewer: What are all possible values for `event` and what do they mean?

Reviewer 2: No, event can be `ADD`, `UPDATE` and `DELETE` here.

Reviewer: This is an API to add a memory - in what scenario is the `event` field going to be anything other than `ADD`?

---
layout: default
title: Add agentic memory
parent: Agentic Memory APIs
grand_parent: ML Commons APIs
nav_order: 40
---

# Add Agentic Memory API

> **Introduced 3.2**

Reviewer 3: We should mark this as experimental

Reviewer: It is marked as experimental on line 13

{: .warning}

Use this API to add an agentic memory to a [memory container]({{site.url}}{{site.baseurl}}/ml-commons-plugin/api/agentic-memory-apis/create-memory-container). You can create a memory in one of the following modes (controlled by the `infer` parameter):

- Fact memory -- A processed representation of the message. The large language model (LLM) associated with the memory container extracts and stores key factual information or knowledge from the original text.

- Raw message memory -- The unprocessed message content.

Once an agentic memory is created, you'll provide its `memory_id` to other APIs.

## Endpoint

```json
POST /_plugins/_ml/memory_containers/{memory_container_id}/memories
```

## Request body fields

The following table lists the available request body fields.

Field | Data type | Required/Optional | Description
:--- | :--- | :--- | :---
`messages` | List | Required | A list of messages.
`session_id` | String | Optional | The session ID associated with the memory.
`agent_id` | String | Optional | The agent ID associated with the memory.

> `infer` | Boolean | Optional | Controls whether the LLM infers context from messages. Default is `true`. When `true`, the LLM extracts factual information from the original text and stores it as the memory. When `false`, the memory contains the unprocessed message and you must explicitly specify the `role` in each message. 

Reviewer: Can this `role` be an arbitrary string? How does the user know what to specify as a `role`?

    ],
    "session_id": "sess_789",
    "agent_id": "agent_123",
    "tags": {
        "topic": "personal info"
    }
}
```
{% include copy-curl.html %}

## Example response

```json
{
    "results": [
        {
            "id": "T9jtmpgBOh0h20Y91WtZ",
            "text": "Machine learning is a subset of artificial intelligence",
            "event": "ADD"
        }
    ],
    "session_id": "sess_789"
}
```

## Response body fields

The following table lists all response body fields.

| Field           | Data type | Description                                                                                       |
| :-------------- | :-------- | :------------------------------------------------------------------------------------------------ |
| `results`       | List      | A list of memory entries returned by the request.                                                 |
| `results.id`    | String    | The unique identifier for the memory entry.                                                       |
| `results.text`  | String    | If `infer` is `false`, contains the stored text from the message. If `infer` is `true`, contains the extracted fact from the message.             |

> | `results.event` | String    | The type of event for the memory entry. For the Add Agentic Memory API, `ADD` indicates that the memory was added. The type of event will always be `ADD`. |

Reviewer: ```suggestion
| `results.event` | String    | The type of event for the memory entry. For the Add Agentic Memory API, the event type is always `ADD`, indicating that the memory was added. |
```

Reviewer 3: Please update this as ` event can be ADD, UPDATE and DELETE here.`

Field | Data type | Required/Optional | Description
:--- | :--- | :--- | :---

> `messages` | List | Required | The list of messages, currently limited to one message per request.

Reviewer 2: Currently we support multiple messages actually. Additionally please add explanation on that user need to explicitly specify role if `infer=false`
